ruby/ext/google/protobuf_c/map.c

```diff
@@ -0,0 +1,894 @@
+      switch (self->key_type) {
+        case UPB_TYPE_BOOL:
+          if (key != Qtrue && key != Qfalse) {
+            rb_raise(rb_eTypeError, "Key must be true or false");
+          }
+          key_val = (key == Qtrue) ? 1 : 0;
+          break;
+        case UPB_TYPE_INT32:
+          native_slot_check_int_range_precision(self->key_type, key);
+          key_val = NUM2INT(key);
+          break;
+        case UPB_TYPE_INT64:
+          native_slot_check_int_range_precision(self->key_type, key);
+          key_val = NUM2LL(key);
+          break;
+        default:
+          break;
+      }
+
+      int64_t* int64_key = (int64_t*)buf;
+      *int64_key = key_val;
+      *out_key = buf;
+      *out_length = sizeof(int64_t);
+      break;
+    }
+
+    case UPB_TYPE_UINT32:
+    case UPB_TYPE_UINT64: {
+      // Unsigned numeric types: use a uint64 in host-native byte order.
+      uint64_t key_val = 0;
+
+      // Do a range/value check.
+      native_slot_check_int_range_precision(self->key_type, key);
+      switch (self->key_type) {
+        case UPB_TYPE_UINT32:
+          key_val = NUM2UINT(key);
+          break;
+        case UPB_TYPE_UINT64:
+          key_val = NUM2ULL(key);
+          break;
+        default:
+          break;
+      }
+
+      uint64_t* uint64_key = (uint64_t*)buf;
+      *uint64_key = key_val;
+      *out_key = buf;
+      *out_length = sizeof(uint64_t);
+      break;
+    }
+
+    default:
+      // Map constructor should not allow a Map with another key type to be
+      // constructed.
+      assert(false);
+      break;
+  }
+}
+
+static VALUE table_key_to_ruby(Map* self, const char* buf, size_t length) {
+  switch (self->key_type) {
+    case UPB_TYPE_BYTES:
+    case UPB_TYPE_STRING: {
+      VALUE ret = rb_str_new(buf, length);
+      rb_enc_associate(ret,
+                       (self->key_type == UPB_TYPE_BYTES) ?
+                       kRubyString8bitEncoding : kRubyStringUtf8Encoding);
+      return ret;
+    }
+
+    case UPB_TYPE_BOOL:
+    case UPB_TYPE_INT32:
+    case UPB_TYPE_INT64: {
+      assert(length == sizeof(int64_t));
+      int64_t* int64_key = (int64_t*)buf;
+
+      if (self->key_type == UPB_TYPE_BOOL) {
+        return *int64_key ? Qtrue : Qfalse;
+      } else {
+        return LL2NUM(*int64_key);
+      }
+    }
+
+    case UPB_TYPE_UINT32:
+    case UPB_TYPE_UINT64: {
+      assert(length == sizeof(uint64_t));
+      uint64_t* uint64_key = (uint64_t*)buf;
+      return ULL2NUM(*uint64_key);
+    }
+
+    default:
+      assert(false);
+      return Qnil;
+  }
+}
+
+static upb_ctype_t upb_table_value_type(upb_fieldtype_t value_type) {
+  switch (value_type) {
+    case UPB_TYPE_BOOL:
+    case UPB_TYPE_INT32:
+    case UPB_TYPE_INT64:
+    case UPB_TYPE_UINT32:
+    case UPB_TYPE_UINT64:
+    case UPB_TYPE_ENUM:
+    case UPB_TYPE_FLOAT:
+    case UPB_TYPE_DOUBLE:
+    case UPB_TYPE_STRING:
+    case UPB_TYPE_BYTES:
+    case UPB_TYPE_MESSAGE:
+      return UPB_CTYPE_UINT64;
```

Comment:
Trying to be overly-generic, I suppose -- removed and replaced with the constant at its two call-sites.